### PR TITLE
Feature deflation

### PR DESCRIPTION
**Pull request by asgeirbirkis**
This PR introduces deflation in Chebfun, as discussed in [1] and [2]. For a simple example using deflation,

```
L = 10;
d = [0 L];
N = chebop(@(x,u) diff(u,2)-u.^2+x, d);
 N.lbc = 0;
 N.rbc = sqrt(L);
 r0 = N\0; % First solution
Ndef = deflate(N, r0, 3, .1);   % Deflate
 r1 = Ndef\0; 
```

[1] Numerical Solution of Nonlinear Boundary Value Problems for Ordinary Differential Equations in the Continuous Framework, Asgeir Birkisson, DPhil Thesis

[2] Deflation techniques for finding distinct solutions of nonlinear partial differential equations (P. E. Farrell, Á. Birkisson, S. W. Funke), In SIAM Journal on Scientific Computing, volume 37, 2015.

### CODE REVIEW

File: @chebmatrix/deflationFun.m [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+% See http://www.chebfun.org/ for Chebfun information.
+
+% By default, apply L2 norm deflation
+if ( nargin < 5 )
+    type = 'L2';
+end
+
+% Extract the blocks from the CHEBMATRIX R
+rBlocks = r.blocks;
+
```
Comment (nickhale) [inline @chebmatrix/deflationFun.m]: Please supply some description here of what is going on mathematically. (Directions to an appropriate equations in the paper would be fine.)

File: @chebop/deflate.m [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    % Ensure R is a CHEBMATRIX to pass to DEFLATIONFUN below:
+    r = chebmatrix(r);
+end
+
+if (nargin(N.op) < 2 )
+    error('must pass x and u to N.op');
+end
+
+% Modify the output operator to reflect the deflation:
+N.op = @(x,u) deflationFun(N.op(x,u), u, r, p, alp, type);
```
Comment (nickhale) [inline @chebop/deflate.m]: Are only scalar problems supported?

File: @adchebfun/adchebfun.m [hunk 3]
```diff
@@ -504,6 +504,79 @@
...
+            end
+            
+            % Initalise a variable for the derivative of the deflated operator:
+            defDeriv = 0;
+            
+            % Compute the norms between the function and the roots. At the same
+            % time, build up the derivative of the deflation part.
+            phiVec = zeros(length(r), 1);
+            
+            if ( strcmp(type, 'L2') )
```
Comment (nickhale) [inline @adchebfun/adchebfun.m]: Again, a pointer here to the relevant equations in the paper would be helpful.

File: @chebmatrix/deflationFun.m (hunk 1 above)
Comment (asgeirbirkis) [inline @chebmatrix/deflationFun.m]: Thanks @nickhale for pointing out the references were missing, I added them in 4a08c9d (same for your comment below).

File: @chebop/deflate.m (hunk 2 above)
Comment (asgeirbirkis) [inline @chebop/deflate.m]: @nickhale Currently yes, d618e11 makes it clearer in the documentation that tha's the case. Again, thanks for pointing out it should be better documented!

**Pull request merged**